File: nexago_app/lib/core/search/search_keywords.dart
/// Normalização de termos de busca (espelho de `functions/src/search-keywords.ts`).
///
/// O app GRAVA `keywords` no perfil (`athlete_profile_repository`), então este
/// arquivo e o canônico em `functions/` precisam gerar o MESMO conjunto — se
/// divergirem, salvar o perfil pelo app rebaixa a busca do atleta.
library;

const int kSearchMinPrefixLength = 2;
const int kSearchMaxKeywords = 400;

String normalizeSearchTerm(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return '';
  return _stripDiacritics(trimmed)
      .toLowerCase()
      .replaceAll(RegExp(r'[^a-z0-9]'), '');
}

/// Mesma normalização, mas PRESERVANDO o acento (só derruba caixa e pontuação).
///
/// As variantes acentuadas entram em `keywords` como apólice: qualquer
/// superfície que consulte sem passar por [normalizeSearchTerm] ainda acha o
/// atleta. Só é emitida quando difere da forma sem acento.
String normalizeAccentedTerm(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return '';
  return trimmed
      .toLowerCase()
      .replaceAll(RegExp(r'[^\p{L}\p{N}]', unicode: true), '');
}

bool isSearchTermLongEnough(String raw) {
  return normalizeSearchTerm(raw).length >= kSearchMinPrefixLength;
}

/// Remove `@` inicial de apelidos antes de tokenizar.
String normalizeNicknameForSearch(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return '';
  return trimmed.startsWith('@') ? trimmed.substring(1).trim() : trimmed;
}

// Só os separadores que de fato quebram palavra num nome. Apóstrofo e
// parênteses ficam de FORA de propósito: `D'Ávila` tem que virar `davila`
// (a normalização os apaga e cola as partes), não `d` + `avila`.
final RegExp _tokenSeparators = RegExp(r'[\s@._-]+');

Iterable<String> _splitParts(String value) {
  return value.split(_tokenSeparators).where((p) => p.trim().isNotEmpty);
}

bool _looksLikeEmail(String raw) => raw.trim().indexOf('@') > 0;

/// Quebra texto em tokens: palavras, partes de e-mail e separadores `. _ -`.
List<String> tokenizeSearchText(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return [];

  final tokens = <String>[];
  final seen = <String>{};

  void addParts(String value) {
    for (final part in _splitParts(value)) {
      final normalized = normalizeSearchTerm(part);
      if (normalized.isEmpty || !seen.add(normalized)) continue;
      tokens.add(normalized);
    }
  }

  if (trimmed.contains('@')) {
    final atIndex = trimmed.indexOf('@');
    addParts(trimmed.substring(0, atIndex));
    addParts(trimmed.substring(atIndex + 1));
  } else {
    addParts(trimmed);
  }

  return tokens;
}

/// Tokens preservando acento, na mesma quebra de [tokenizeSearchText].
List<String> _tokenizeAccentedText(String raw) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) return [];
  final atIndex = trimmed.indexOf('@');
  final source = atIndex >= 0
      ? '${trimmed.substring(0, atIndex)} ${trimmed.substring(atIndex + 1)}'
      : trimmed;

  final out = <String>[];
  final seen = <String>{};
  for (final part in _splitParts(source)) {
    final accented = normalizeAccentedTerm(part);
    if (accented.isEmpty || !seen.add(accented)) continue;
    out.add(accented);
  }
  return out;
}

/// Todas as formas indexáveis de um texto: cada palavra sem acento, cada
/// palavra com acento (quando difere) e a forma COLADA do texto inteiro — é
/// ela que faz `ana_paula` e `João Silva` casarem com quem digita `anapaula` /
/// `joaosilva`.
List<String> searchVariants(String raw) {
  final ascii = tokenizeSearchText(raw);
  if (ascii.isEmpty) return const [];

  final out = <String>[...ascii];
  final seen = <String>{...ascii};

  void push(String value) {
    if (value.isEmpty || !seen.add(value)) return;
    out.add(value);
  }

  // A forma colada é para nome/apelido. E-mail fica de fora: `maria@gmail.com`
  // viraria `mariagmailcom`, dezenas de prefixos sem valor de busca — e o
  // endereço reconstruído num espelho que é declaradamente sem PII.
  if (ascii.length > 1 && !_looksLikeEmail(raw)) push(ascii.join());
  for (final accented in _tokenizeAccentedText(raw)) {
    push(accented);
  }

  return out;
}

List<String> _wordPrefixes(String token, int minPrefix, int maxLen) {
  if (token.isEmpty) return const [];

  final limit = token.length < maxLen ? token.length : maxLen;
  final out = <String>{};
  final start = minPrefix > limit ? limit : (minPrefix < 1 ? 1 : minPrefix);
  for (var i = start; i <= limit; i++) {
    out.add(token.substring(0, i));
  }
  out.add(token);
  return out.toList();
}

/// Gera as `keywords` de busca a partir de várias fontes de texto.
///
/// Duas passadas de propósito: primeiro a forma EXATA de toda variante, depois
/// os prefixos. Assim, se o teto de [maxKeywords] estourar num nome muito
/// longo, o apelido ainda tem sua palavra inteira indexada — a passada única
/// gastava o orçamento todo nos prefixos da primeira fonte.
List<String> generateKeywords(
  List<String> sources, {
  int minPrefix = kSearchMinPrefixLength,
  int maxKeywords = kSearchMaxKeywords,
}) {
  final variants = <String>[];
  final seenVariants = <String>{};
  for (final source in sources) {
    if (source.trim().isEmpty) continue;
    for (final variant in searchVariants(source)) {
      if (seenVariants.add(variant)) variants.add(variant);
    }
  }

  final keywords = <String>{};
  for (final variant in variants) {
    if (keywords.length >= maxKeywords) break;
    keywords.add(variant);
  }
  for (final variant in variants) {
    if (keywords.length >= maxKeywords) break;
    for (final prefix in _wordPrefixes(variant, minPrefix, 32)) {
      keywords.add(prefix);
      if (keywords.length >= maxKeywords) break;
    }
  }

  return keywords.toList()..sort();
}

/// Tokens do termo digitado, na ordem em que foram escritos.
List<String> searchQueryTokens(String raw) => tokenizeSearchText(raw);

/// Token que vai no `array-contains`: o mais longo é o mais seletivo, então
/// `de oliveira` consulta por `oliveira` e não pela preposição. Os demais
/// tokens viram filtro no client ([profileMatchesSearchTokens]).
String searchAnchorToken(List<String> tokens) {
  var anchor = '';
  for (final token in tokens) {
    if (token.length > anchor.length) anchor = token;
  }
  return anchor;
}

/// Texto de um perfil que o filtro de busca enxerga.
class SearchableProfileText {
  const SearchableProfileText({
    this.fullName,
    this.nickname,
    this.keywords = const [],
  });

  final String? fullName;
  final String? nickname;
  final List<String> keywords;

  List<String> get _texts {
    final out = <String>[];
    final name = fullName?.trim();
    if (name != null && name.isNotEmpty) out.add(name);
    final nick = normalizeNicknameForSearch(nickname ?? '');
    if (nick.isNotEmpty) out.add(nick);
    return out;
  }

  List<String> get variants =>
      _texts.expand(searchVariants).toList(growable: false);

  List<String> get words =>
      _texts.expand(tokenizeSearchText).toList(growable: false);
}

/// O documento casa com TODOS os tokens digitados? É o `AND` que o Firestore
/// não faz: `array-contains` aceita um valor só, então a consulta ancora no
/// token mais seletivo e este filtro corta o resto — sem leitura extra, porque
/// `keywords` e o nome já vieram no próprio doc.
bool profileMatchesSearchTokens(
  SearchableProfileText profile,
  List<String> tokens,
) {
  if (tokens.isEmpty) return false;
  final variants = profile.variants;
  return tokens.every((token) {
    if (profile.keywords.contains(token)) return true;
    // Confere também contra o nome do próprio doc: `keywords` gravado por uma
    // versão antiga do gerador não pode esconder um atleta que casa de fato.
    return variants.any((variant) => variant.startsWith(token));
  });
}

/// Ordem de exibição do resultado: quanto menor, mais relevante. Sem isso o
/// `limit` do Firestore devolve um recorte arbitrário e o casamento exato pode
/// ficar de fora da página.
int searchRelevanceScore(
  SearchableProfileText profile,
  List<String> tokens,
) {
  if (tokens.isEmpty) return 9;
  final joined = tokens.join();
  final nickname = normalizeSearchTerm(
    normalizeNicknameForSearch(profile.nickname ?? ''),
  );
  final fullName = normalizeSearchTerm(profile.fullName?.trim() ?? '');
  final words = profile.words;

  if (nickname == joined || fullName == joined) return 0;
  if (words.contains(joined)) return 1;
  if (nickname.startsWith(joined) || fullName.startsWith(joined)) return 2;
  if (words.any((word) => word.startsWith(tokens.first))) return 3;
  return 4;
}

class UserSearchFields {
  const UserSearchFields({
    required this.keywords,
    required this.hasAthleteRole,
    required this.hasOrganizerRole,
  });

  final List<String> keywords;
  final bool hasAthleteRole;
  final bool hasOrganizerRole;
}

bool _userDocHasRole(Map<String, dynamic> data, String role) {
  final roles = data['roles'];
  return roles is List &&
      roles.any((r) => r is String && r.trim().toLowerCase() == role);
}

String _readString(Map<String, dynamic> data, String key) {
  final v = data[key];
  return v is String ? v.trim() : '';
}

/// Monta `keywords` e flags de papel a partir de um mapa de usuário Firestore.
UserSearchFields buildUserSearchFields(Map<String, dynamic> data) {
  final fullName = _readString(data, 'fullName').isNotEmpty
      ? _readString(data, 'fullName')
      : _readString(data, 'name');
  final nickname = normalizeNicknameForSearch(_readString(data, 'nickname'));
  final email = _readString(data, 'email');

  return UserSearchFields(
    keywords: generateKeywords([fullName, nickname, email]),
    hasAthleteRole: _userDocHasRole(data, 'athlete'),
    hasOrganizerRole: _userDocHasRole(data, 'organizer'),
  );
}

String _stripDiacritics(String value) {
  const accents = 'àáâãäåèéêëìíîïòóôõöùúûüçñýÿ';
  const plain = 'aaaaaaeeeeiiiiooooouuuucnyy';
  final buffer = StringBuffer();
  for (final rune in value.runes) {
    final char = String.fromCharCode(rune);
    final lower = char.toLowerCase();
    final idx = accents.indexOf(lower);
    if (idx >= 0) {
      buffer.write(plain[idx]);
    } else {
      buffer.write(char);
    }
  }
  return buffer.toString();
}

/// Prefixos para busca case-insensitive de `nickname` no Firestore.
List<String> nicknameSearchPrefixes(String raw) {
  final v = raw.trim();
  if (v.isEmpty) return const [];
  if (v.length == 1) {
    return [v, v.toLowerCase(), v.toUpperCase()];
  }
  final lower = v.toLowerCase();
  final title = '${v[0].toUpperCase()}${v.substring(1).toLowerCase()}';
  return {v, lower, v.toUpperCase(), title}.toList();
}

File: functions/src/search-keywords.ts
/** Prefixos de palavras para busca Firestore (`keywords` + `array-contains`). */

export const DEFAULT_MIN_PREFIX = 2;
export const DEFAULT_MAX_KEYWORDS = 400;

export type GenerateKeywordsOptions = {
  minPrefix?: number;
  maxKeywords?: number;
};

export type UserSearchFields = {
  keywords: string[];
  hasAthleteRole: boolean;
  hasOrganizerRole: boolean;
};

export type TournamentSearchFields = {
  keywords: string[];
};

export type LeagueSearchFields = {
  keywords: string[];
};

export type TeamSearchFields = {
  keywords: string[];
  player1DisplayName?: string;
  player2DisplayName?: string;
};

function stripDiacritics(value: string): string {
  return value.normalize("NFD").replace(/\p{M}/gu, "");
}

/** Normaliza termo digitado na busca (lowercase, sem acentos, alfanumérico). */
export function normalizeSearchTerm(raw: string): string {
  const trimmed = raw.trim();
  if (!trimmed) return "";
  return stripDiacritics(trimmed)
    .toLowerCase()
    .replace(/[^a-z0-9]/g, "");
}

/**
 * Mesma normalização, mas PRESERVANDO o acento (só derruba caixa e pontuação).
 * As variantes acentuadas entram em `keywords` como apólice: qualquer
 * superfície que consulte sem passar por `normalizeSearchTerm` ainda acha o
 * atleta. Só é emitida quando difere da forma sem acento.
 */
export function normalizeAccentedTerm(raw: string): string {
  const trimmed = raw.trim();
  if (!trimmed) return "";
  return trimmed
    .normalize("NFC")
    .toLowerCase()
    .replace(/[^\p{L}\p{N}]/gu, "");
}

function normalizeToken(raw: string): string {
  return normalizeSearchTerm(raw);
}

// Só os separadores que de fato quebram palavra num nome. Apóstrofo e
// parênteses ficam de FORA de propósito: `D'Ávila` tem que virar `davila`
// (a normalização os apaga e cola as partes), não `d` + `avila`.
const TOKEN_SEPARATORS = /[\s@._-]+/;

function splitParts(value: string): string[] {
  return value.split(TOKEN_SEPARATORS).filter((p) => p.trim().length > 0);
}

/** Quebra texto em tokens: palavras, partes de e-mail e separadores `. _ -`. */
export function tokenizeSearchText(raw: string): string[] {
  const trimmed = raw.trim();
  if (!trimmed) return [];

  const tokens: string[] = [];
  const seen = new Set<string>();
  const addParts = (value: string) => {
    for (const part of splitParts(value)) {
      const normalized = normalizeToken(part);
      if (!normalized || seen.has(normalized)) continue;
      seen.add(normalized);
      tokens.push(normalized);
    }
  };

  if (trimmed.includes("@")) {
    const atIndex = trimmed.indexOf("@");
    addParts(trimmed.slice(0, atIndex));
    addParts(trimmed.slice(atIndex + 1));
  } else {
    addParts(trimmed);
  }

  return tokens;
}

/** Tokens preservando acento, na mesma quebra de `tokenizeSearchText`. */
function tokenizeAccentedText(raw: string): string[] {
  const trimmed = raw.trim();
  if (!trimmed) return [];
  const source = trimmed.includes("@") ?
    `${trimmed.slice(0, trimmed.indexOf("@"))} ${trimmed.slice(
      trimmed.indexOf("@") + 1
    )}` :
    trimmed;
  const out: string[] = [];
  const seen = new Set<string>();
  for (const part of splitParts(source)) {
    const accented = normalizeAccentedTerm(part);
    if (!accented || seen.has(accented)) continue;
    seen.add(accented);
    out.push(accented);
  }
  return out;
}

function looksLikeEmail(raw: string): boolean {
  const at = raw.trim().indexOf("@");
  return at > 0;
}

/**
 * Todas as formas indexáveis de um texto: cada palavra sem acento, cada
 * palavra com acento (quando difere) e a forma COLADA do texto inteiro — é
 * ela que faz `"ana_paula"` e `"João Silva"` casarem com quem digita
 * `"anapaula"` / `"joaosilva"`.
 */
export function searchVariants(raw: string): string[] {
  const ascii = tokenizeSearchText(raw);
  if (ascii.length === 0) return [];

  const out: string[] = [...ascii];
  const seen = new Set(ascii);

  const push = (value: string) => {
    if (!value || seen.has(value)) return;
    seen.add(value);
    out.push(value);
  };

  // A forma colada é para nome/apelido (`ana_paula` → `anapaula`). E-mail
  // fica de fora: `maria@gmail.com` viraria `mariagmailcom`, dezenas de
  // prefixos sem valor de busca — e o endereço reconstruído num espelho que
  // é declaradamente sem PII.
  if (ascii.length > 1 && !looksLikeEmail(raw)) push(ascii.join(""));
  for (const accented of tokenizeAccentedText(raw)) push(accented);

  return out;
}

function wordPrefixes(
  token: string,
  minPrefix: number,
  maxLen: number
): string[] {
  if (!token) return [];

  const limit = Math.min(token.length, maxLen);
  const out = new Set<string>();
  const start = Math.max(1, Math.min(minPrefix, limit));
  for (let i = start; i <= limit; i++) {
    out.add(token.slice(0, i));
  }
  out.add(token);
  return [...out];
}

/**
 * Gera as `keywords` de busca a partir de várias fontes de texto.
 *
 * Duas passadas de propósito: primeiro a forma EXATA de toda variante, depois
 * os prefixos. Assim, se o teto de `maxKeywords` estourar num nome muito
 * longo, o apelido ainda tem sua palavra inteira indexada — a passada única
 * gastava o orçamento todo nos prefixos da primeira fonte.
 */
export function generateKeywords(
  sources: string[],
  options: GenerateKeywordsOptions = {}
): string[] {
  const minPrefix = options.minPrefix ?? DEFAULT_MIN_PREFIX;
  const maxKeywords = options.maxKeywords ?? DEFAULT_MAX_KEYWORDS;

  const variants: string[] = [];
  const seenVariants = new Set<string>();
  for (const source of sources) {
    if (!source?.trim()) continue;
    for (const variant of searchVariants(source)) {
      if (seenVariants.has(variant)) continue;
      seenVariants.add(variant);
      variants.push(variant);
    }
  }

  const keywords = new Set<string>();
  for (const variant of variants) {
    if (keywords.size >= maxKeywords) break;
    keywords.add(variant);
  }
  for (const variant of variants) {
    if (keywords.size >= maxKeywords) break;
    for (const prefix of wordPrefixes(variant, minPrefix, 32)) {
      keywords.add(prefix);
      if (keywords.size >= maxKeywords) break;
    }
  }

  return [...keywords].sort();
}

/** Tokens do termo digitado, na ordem em que foram escritos. */
export function searchQueryTokens(raw: string): string[] {
  return tokenizeSearchText(raw);
}

/**
 * Token que vai no `array-contains`: o mais longo é o mais seletivo, então
 * `"de oliveira"` consulta por `oliveira` e não pela preposição. Os demais
 * tokens viram filtro no client (`profileMatchesSearchTokens`).
 */
export function searchAnchorToken(tokens: readonly string[]): string {
  let anchor = "";
  for (const token of tokens) {
    if (token.length > anchor.length) anchor = token;
  }
  return anchor;
}

export type SearchableProfile = {
  fullName?: string | null;
  name?: string | null;
  nickname?: string | null;
  keywords?: readonly string[] | null;
};

function profileTexts(profile: SearchableProfile): string[] {
  const out: string[] = [];
  const push = (v: string | null | undefined) => {
    if (typeof v === "string" && v.trim()) out.push(v);
  };
  push(profile.fullName);
  push(profile.name);
  push(profile.nickname ? normalizeNicknameForSearch(profile.nickname) : "");
  return out;
}

function tokenMatchesProfile(
  profile: SearchableProfile,
  variants: readonly string[],
  token: string
): boolean {
  if (Array.isArray(profile.keywords) && profile.keywords.includes(token)) {
    return true;
  }
  // Confere também contra o nome do próprio doc: `keywords` gravado por uma
  // versão antiga do gerador não pode esconder um atleta que casa de fato.
  return variants.some((variant) => variant.startsWith(token));
}

/**
 * O documento casa com TODOS os tokens digitados? É o `AND` que o Firestore
 * não faz: `array-contains` aceita um valor só, então a consulta ancora no
 * token mais seletivo e este filtro corta o resto — sem leitura extra, porque
 * `keywords` e o nome já vieram no próprio doc.
 */
export function profileMatchesSearchTokens(
  profile: SearchableProfile,
  tokens: readonly string[]
): boolean {
  if (tokens.length === 0) return false;
  const variants = profileTexts(profile).flatMap((text) => searchVariants(text));
  return tokens.every((token) => tokenMatchesProfile(profile, variants, token));
}

/**
 * Ordem de exibição do resultado: quanto menor, mais relevante. Sem isso o
 * `limit` do Firestore devolve um recorte arbitrário e o casamento exato pode
 * ficar de fora da página.
 */
export function searchRelevanceScore(
  profile: SearchableProfile,
  tokens: readonly string[]
): number {
  if (tokens.length === 0) return 9;
  const joined = tokens.join("");
  const nickname = normalizeSearchTerm(
    normalizeNicknameForSearch(profile.nickname ?? "")
  );
  const fullName = normalizeSearchTerm(
    profile.fullName?.trim() || profile.name?.trim() || ""
  );
  const words = profileTexts(profile).flatMap((text) =>
    tokenizeSearchText(text)
  );

  if (nickname === joined || fullName === joined) return 0;
  if (words.includes(joined)) return 1;
  if (nickname.startsWith(joined) || fullName.startsWith(joined)) return 2;
  if (words.some((word) => word.startsWith(tokens[0] ?? ""))) return 3;
  return 4;
}

function userDocHasRole(data: Record<string, unknown>, role: string): boolean {
  const roles = data.roles;
  return Array.isArray(roles) &&
    roles.some((r) => typeof r === "string" && r.trim().toLowerCase() === role);
}

function readString(data: Record<string, unknown>, key: string): string {
  const v = data[key];
  return typeof v === "string" ? v.trim() : "";
}

/** Remove `@` inicial de apelidos antes de tokenizar. */
export function normalizeNicknameForSearch(raw: string): string {
  const trimmed = raw.trim();
  if (!trimmed) return "";
  return trimmed.startsWith("@") ? trimmed.slice(1).trim() : trimmed;
}

export function buildUserSearchFields(
  data: Record<string, unknown>
): UserSearchFields {
  const fullName = readString(data, "fullName") || readString(data, "name");
  const nickname = normalizeNicknameForSearch(readString(data, "nickname"));
  const email = readString(data, "email");

  return {
    keywords: generateKeywords([fullName, nickname, email]),
    hasAthleteRole: userDocHasRole(data, "athlete"),
    hasOrganizerRole: userDocHasRole(data, "organizer"),
  };
}

export function buildTournamentSearchFields(
  data: Record<string, unknown>
): TournamentSearchFields {
  const name = readString(data, "name");
  const city = readString(data, "city");
  const location = readString(data, "location");
  const seasonLabel = readString(data, "seasonLabel");

  return {
    keywords: generateKeywords([name, city, location, seasonLabel]),
  };
}

export function buildLeagueSearchFields(
  data: Record<string, unknown>
): LeagueSearchFields {
  const name = readString(data, "name");
  const city = readString(data, "city");
  const seasonLabel =
    readString(data, "seasonLabel") || readString(data, "season");
  const sources = [name, city, seasonLabel];

  const stages = data.stages;
  if (Array.isArray(stages)) {
    for (const stage of stages) {
      if (!stage || typeof stage !== "object") continue;
      const stageName = readString(stage as Record<string, unknown>, "name");
      if (stageName) sources.push(stageName);
    }
  }

  return {
    keywords: generateKeywords(sources),
  };
}

export function buildTeamSearchFields(
  data: Record<string, unknown>,
  playerNames: string[] = []
): TeamSearchFields {
  const teamName = readString(data, "teamName");
  const p1 = readString(data, "player1DisplayName");
  const p2 = readString(data, "player2DisplayName");
  const sources = [teamName, p1, p2, ...playerNames].filter((s) => s.length > 0);

  const fields: TeamSearchFields = {
    keywords: generateKeywords(sources),
  };

  const resolvedP1 = p1 || playerNames[0] || "";
  const resolvedP2 = p2 || playerNames[1] || "";
  if (resolvedP1) fields.player1DisplayName = resolvedP1;
  if (resolvedP2) fields.player2DisplayName = resolvedP2;

  return fields;
}

export function searchFieldsChanged(
  current: Record<string, unknown> | undefined,
  next: Record<string, unknown>
): boolean {
  if (!current) return true;

  for (const [key, value] of Object.entries(next)) {
    const prev = current[key];
    if (Array.isArray(value) && Array.isArray(prev)) {
      if (value.length !== prev.length) return true;
      const a = [...value].map(String).sort();
      const b = [...prev].map(String).sort();
      for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) return true;
      }
      continue;
    }
    if (prev !== value) return true;
  }
  return false;
}

export function userSearchSourceFieldsChanged(
  before: Record<string, unknown> | undefined,
  after: Record<string, unknown> | undefined
): boolean {
  if (!after) return false;
  if (!before) return true;

  const keys = ["fullName", "name", "nickname", "email", "roles"];
  for (const key of keys) {
    const b = before[key];
    const a = after[key];
    if (JSON.stringify(b) !== JSON.stringify(a)) return true;
  }
  return false;
}

export function tournamentSearchSourceFieldsChanged(
  before: Record<string, unknown> | undefined,
  after: Record<string, unknown> | undefined
): boolean {
  if (!after) return false;
  if (!before) return true;

  const keys = ["name", "city", "location", "seasonLabel"];
  for (const key of keys) {
    if (before[key] !== after[key]) return true;
  }
  return false;
}

export function leagueSearchSourceFieldsChanged(
  before: Record<string, unknown> | undefined,
  after: Record<string, unknown> | undefined
): boolean {
  if (!after) return false;
  if (!before) return true;

  const keys = ["name", "city", "seasonLabel", "season", "stages"];
  for (const key of keys) {
    if (JSON.stringify(before[key]) !== JSON.stringify(after[key])) return true;
  }
  return false;
}

export function teamSearchSourceFieldsChanged(
  before: Record<string, unknown> | undefined,
  after: Record<string, unknown> | undefined
): boolean {
  if (!after) return false;
  if (!before) return true;

  const keys = [
    "teamName",
    "player1Id",
    "player2Id",
    "player1DisplayName",
    "player2DisplayName",
  ];
  for (const key of keys) {
    if (before[key] !== after[key]) return true;
  }
  return false;
}
